File: ios/Sources/UnifiedTrackingPlugin/Analytics/MixpanelProvider.swift
import Foundation
import Mixpanel

/// Mixpanel (mixpanel-swift) bridge for the unified-tracking plugin.
///
/// Reads `token` from the provider config and initializes a `MixpanelInstance`.
/// If no token is supplied the SDK is not started and the provider stays
/// uninitialized (so ProviderManager will skip it).
///
/// NOTE: The host app does NOT need to call `Mixpanel.initialize(...)` on its
/// own — this provider handles initialization. Calling it twice with the same
/// token is safe (Mixpanel returns the existing instance).
class MixpanelProvider: AnalyticsProvider {
    let name = "mixpanel"
    private(set) var isEnabled: Bool
    private(set) var isInitialized: Bool

    private let config: [String: Any]
    private var debugMode = false
    /// Tracks the consent state for parity. Mixpanel exposes
    /// `optInTracking()` / `optOutTracking()` which we wire in `setConsent`.
    private var consentGranted = true
    /// The Mixpanel instance returned by `Mixpanel.initialize(...)`. Held weakly-
    /// typed as `MixpanelInstance?` so every method can guard on it.
    private var instance: MixpanelInstance?

    init(config: [String: Any]) {
        self.config = config
        self.isEnabled = config["enabled"] as? Bool ?? true
        self.isInitialized = false

        guard let token = config["token"] as? String, !token.isEmpty else {
            // No token -> cannot start Mixpanel. Leave uninitialized.
            print("[MixpanelProvider] No token provided; Mixpanel not started.")
            return
        }

        // NOTE(unverified): build with Xcode — confirm
        // `Mixpanel.initialize(token:trackAutomaticEvents:)` (or `Mixpanel.initialize(token:)`)
        // signature in mixpanel-swift ~4.2.8. `trackAutomaticEvents` became a required
        // parameter in recent versions; passing `false` keeps behaviour predictable.
        let trackAutomatic = config["trackAutomaticEvents"] as? Bool ?? false
        self.instance = Mixpanel.initialize(token: token, trackAutomaticEvents: trackAutomatic)

        self.isInitialized = true
    }

    func trackEvent(_ event: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `track(event:properties:)` accepts
        // `Properties` (= `[String: MixpanelType]`) in mixpanel-swift ~4.2.8.
        instance.track(event: event, properties: Self.sanitizeProperties(properties))
    }

    func identifyUser(_ userId: String, traits: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `identify(distinctId:)` and
        // `people.set(properties:)` argument labels in mixpanel-swift ~4.2.8.
        instance.identify(distinctId: userId)
        if !traits.isEmpty {
            instance.people.set(properties: Self.sanitizeProperties(traits))
        }
    }

    func setUserProperties(_ properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }
        instance.people.set(properties: Self.sanitizeProperties(properties))
    }

    func logRevenue(amount: Double, currency: String, productId: String?, quantity: Int, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }

        var chargeProperties: [String: MixpanelType] = [
            "currency": currency as MixpanelType,
            "quantity": quantity as MixpanelType
        ]
        if let productId = productId {
            chargeProperties["productId"] = productId as MixpanelType
        }
        // Caller-supplied properties take precedence over the derived defaults.
        for (key, value) in Self.sanitizeProperties(properties) {
            chargeProperties[key] = value
        }

        // NOTE(unverified): build with Xcode — confirm `people.trackCharge(amount:properties:)`
        // signature in mixpanel-swift ~4.2.8. The `amount` is a `Double`.
        instance.people.trackCharge(amount: amount, properties: chargeProperties)
    }

    func logScreenView(_ screenName: String, properties: [String: Any]) {
        guard isEnabled && isInitialized && consentGranted, let instance = instance else { return }

        var params: [String: MixpanelType] = [
            "screen": screenName as MixpanelType
        ]
        for (key, value) in Self.sanitizeProperties(properties) {
            params[key] = value
        }

        instance.track(event: "Screen View", properties: params)
    }

    func setConsent(_ granted: Bool) {
        consentGranted = granted
        guard let instance = instance else { return }
        // NOTE(unverified): build with Xcode — confirm `optInTracking()` /
        // `optOutTracking()` exist as instance methods in mixpanel-swift ~4.2.8.
        if granted {
            instance.optInTracking()
        } else {
            instance.optOutTracking()
        }
    }

    func reset() {
        guard let instance = instance else { return }
        instance.reset()
    }

    func setDebugMode(_ enabled: Bool) {
        // Mixpanel's logging level is configured at SDK init time (via
        // `Mixpanel.mainInstance().loggingEnabled = true` on older versions, or
        // `MixpanelInstance.loggingEnabled` on current versions). We retain the
        // flag here for parity with other providers; flipping it at runtime is
        // best-effort.
        debugMode = enabled
        // NOTE(unverified): build with Xcode — confirm `instance.loggingEnabled`
        // is still a writable property in mixpanel-swift ~4.2.8.
        instance?.loggingEnabled = enabled
    }

    // MARK: - Helpers

    /// Mixpanel's `Properties` type is `[String: MixpanelType]`. `MixpanelType` is a
    /// protocol that String, NSNumber, Bool, Int, UInt, Double, Float, [MixpanelType],
    /// [String: MixpanelType], Date, URL, NSNull all conform to. Anything else is
    /// coerced to `String(describing:)` so the SDK never receives an unsupported type.
    private static func sanitizeProperties(_ properties: [String: Any]) -> [String: MixpanelType] {
        var result: [String: MixpanelType] = [:]
        for (key, value) in properties {
            if let typed = value as? MixpanelType {
                result[key] = typed
            } else {
                // Fall back to a string description.
                result[key] = String(describing: value) as MixpanelType
            }
        }
        return result
    }
}

File: android/src/main/java/com/aoneahsan/unifiedtracking/providers/analytics/MixpanelProvider.java
package com.aoneahsan.unifiedtracking.providers.analytics;

import android.content.Context;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.aoneahsan.unifiedtracking.providers.AnalyticsProvider;
import com.mixpanel.android.mpmetrics.MixpanelAPI;

import org.json.JSONObject;

import java.util.Map;

/**
 * Mixpanel analytics provider for Android.
 *
 * Uses {@code com.mixpanel.android.mpmetrics.MixpanelAPI} as the SDK entry point.
 * The SDK is instantiated via {@code MixpanelAPI.getInstance(context, token, trackAutomaticEvents)}
 * which is the documented factory in Mixpanel Android SDK 7.x.
 *
 * NOTE(unverified): built against the Mixpanel Android SDK 7.5.x API. Confirm exact
 * signatures (MixpanelAPI.getInstance overloads, People.set/trackCharge, optInTracking/
 * optOutTracking) in Android Studio against the resolved SDK version.
 */
public class MixpanelProvider implements AnalyticsProvider {
    private static final String TAG = "Mixpanel";
    private final Context context;
    private final JSObject config;
    private MixpanelAPI mixpanel;
    private boolean enabled;
    private boolean initialized;
    private boolean debugMode = false;

    public MixpanelProvider(Context context, JSObject config) {
        this.context = context;
        this.config = config;
        this.enabled = config.getBoolean("enabled", true);
        this.initialized = false;

        try {
            String token = config.getString("token");
            if (token == null || token.trim().isEmpty()) {
                Log.e(TAG, "Mixpanel token is missing; provider will be inert.");
                return;
            }

            // NOTE(unverified): MixpanelAPI.getInstance(Context, String, boolean) is the
            // current 3-arg factory in Mixpanel Android SDK 7.x. The third arg disables
            // automatic event tracking so only events from this provider are reported.
            this.mixpanel = MixpanelAPI.getInstance(context, token, false);
            if (this.mixpanel == null) {
                Log.e(TAG, "MixpanelAPI.getInstance returned null; provider will be inert.");
                return;
            }
            this.initialized = true;

            if (debugMode) {
                Log.d(TAG, "Mixpanel initialized");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize Mixpanel", e);
            this.initialized = false;
        }
    }

    @Override
    public String getName() {
        return "mixpanel";
    }

    @Override
    public boolean isEnabled() {
        return enabled;
    }

    @Override
    public boolean isInitialized() {
        return initialized;
    }

    @Override
    public void trackEvent(String event, Map<String, Object> properties) {
        if (!enabled || !initialized || mixpanel == null || event == null) return;

        try {
            JSONObject props = mapToJson(properties);
            mixpanel.track(event, props);

            if (debugMode) {
                Log.d(TAG, "Tracked event: " + event);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to track event", e);
        }
    }

    @Override
    public void identifyUser(String userId, Map<String, Object> traits) {
        if (!enabled || !initialized || mixpanel == null || userId == null) return;

        try {
            // NOTE(unverified): MixpanelAPI.identify(String) replaced the legacy
            // identify(String, boolean) overload in Mixpanel Android SDK 6.x+. Confirm
            // the no-second-arg form is correct in the resolved SDK.
            mixpanel.identify(userId);

            if (traits != null && !traits.isEmpty()) {
                JSONObject traitsJson = mapToJson(traits);
                mixpanel.getPeople().set(traitsJson);
            }

            if (debugMode) {
                Log.d(TAG, "Identified user: " + userId);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to identify user", e);
        }
    }

    @Override
    public void setUserProperties(Map<String, Object> properties) {
        if (!enabled || !initialized || mixpanel == null || properties == null) return;

        try {
            JSONObject props = mapToJson(properties);
            mixpanel.getPeople().set(props);

            if (debugMode) {
                Log.d(TAG, "Set user properties");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set user properties", e);
        }
    }

    @Override
    public void logRevenue(double amount, String currency, String productId, int quantity, Map<String, Object> properties) {
        if (!enabled || !initialized || mixpanel == null) return;

        try {
            JSONObject revenueProps = new JSONObject();
            if (currency != null) revenueProps.put("currency", currency);
            if (productId != null) revenueProps.put("productId", productId);
            revenueProps.put("quantity", quantity);

            if (properties != null) {
                for (Map.Entry<String, Object> entry : properties.entrySet()) {
                    if (entry.getValue() == null) continue;
                    revenueProps.put(entry.getKey(), entry.getValue());
                }
            }

            // NOTE(unverified): People.trackCharge(double, JSONObject) is the documented
            // revenue API in Mixpanel Android SDK 7.x. Older SDKs used trackCharge with
            // a Map; confirm in Android Studio.
            mixpanel.getPeople().trackCharge(amount, revenueProps);

            if (debugMode) {
                Log.d(TAG, "Logged revenue: " + amount + " " + currency);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log revenue", e);
        }
    }

    @Override
    public void logScreenView(String screenName, Map<String, Object> properties) {
        if (!enabled || !initialized || mixpanel == null || screenName == null) return;

        try {
            JSONObject props = mapToJson(properties);
            props.put("screen", screenName);
            mixpanel.track("Screen View", props);

            if (debugMode) {
                Log.d(TAG, "Logged screen view: " + screenName);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to log screen view", e);
        }
    }

    @Override
    public void setConsent(boolean granted) {
        if (!initialized || mixpanel == null) return;

        try {
            if (granted) {
                // NOTE(unverified): optInTracking() / optOutTracking() are the documented
                // consent toggles in Mixpanel Android SDK 7.x. optInTracking has overloads
                // that take a distinct id + properties; the no-arg form is the simplest.
                mixpanel.optInTracking();
            } else {
                mixpanel.optOutTracking();
            }
            this.enabled = granted;

            if (debugMode) {
                Log.d(TAG, "Set consent: " + granted);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to set consent", e);
        }
    }

    @Override
    public void reset() {
        if (!initialized || mixpanel == null) return;

        try {
            mixpanel.reset();

            if (debugMode) {
                Log.d(TAG, "Reset Mixpanel state");
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to reset", e);
        }
    }

    @Override
    public void setDebugMode(boolean enabled) {
        this.debugMode = enabled;
        if (enabled) {
            Log.d(TAG, "Debug mode enabled for Mixpanel provider");
        }
    }

    private JSONObject mapToJson(Map<String, Object> map) {
        JSONObject json = new JSONObject();
        if (map == null) return json;

        for (Map.Entry<String, Object> entry : map.entrySet()) {
            if (entry.getKey() == null || entry.getValue() == null) continue;
            try {
                json.put(entry.getKey(), entry.getValue());
            } catch (Exception e) {
                // Fall back to string representation if JSONObject.put rejects the value.
                try {
                    json.put(entry.getKey(), String.valueOf(entry.getValue()));
                } catch (Exception ignored) {
                    // Skip unserializable entries.
                }
            }
        }
        return json;
    }
}
